### PR TITLE
[POSTGRES] Experiment r2dbc-pool library

### PR DESCRIPTION
**Pull request by vttranlina**
ref: https://www.mail-archive.com/server-dev@james.apache.org/msg73461.html

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java [hunk 1]
```diff
@@ -83,4 +98,53 @@ public Mono<PostgresqlConnection> resolver(Domain domain) {
...
+            this.connectionPool = new ConnectionPool(poolConfiguration);
+
+            this.mapConnectionLookup = new HashMap<>();
+        }
+
+        @Override
+        public Mono<? extends Connection> resolver(Domain domain) {
+            return Mono.fromCallable(() -> mapConnectionLookup.get(domain.asString()))
+                .switchIfEmpty(createPooledConnection(domain)
+                    .doOnSuccess(connection -> mapConnectionLookup.put(domain.asString(), connection)));
```
Comment (quantranhong1999) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java]: `doOnSuccess` as I know it is async. Any chance that resolved Connection -> execute bussiness query -> release the connection (remove no Map Entry yet) -> then `doOnSuccess` adds the Map Entry (async and could be a bit late).

Then in that case the Map `mapConnectionLookup` would have a entry that points a Domain to a closed connection.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java [hunk 2]
```diff
@@ -83,4 +98,53 @@ public Mono<PostgresqlConnection> resolver(Domain domain) {
...
+        }
+
+        @Override
+        public Mono<Void> release(Domain domain) {
+            return Mono.fromCallable(() -> mapConnectionLookup.get(domain.asString()))
+                .flatMap(release());
```
Comment (quantranhong1999) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java]: should be sth like `mapConnectionLookup.remove`, no?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java [hunk 3]
```diff
@@ -83,4 +98,53 @@ public Mono<PostgresqlConnection> resolver(Domain domain) {
...
+                            .execute())
+                        .thenReturn(pooledConnection));
+        }
+
+        private static Function<Connection, Mono<Void>> release() {
+            return connection -> Mono.from(connection.createStatement("RESET app.current_domain").execute())
+                .doOnSuccess(e -> {
+                    LOGGER.debug("connection has been released, {}", e.hashCode());
+                })
+                .then(Mono.from(connection.close()));
```
Comment (quantranhong1999) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java:142]: Hmm keep disposing and creating connection again for every business query seems not good to me.

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionPoolTest.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                connectionPool.create()
+                    .flatMapMany(connection -> Flux.from(connection.createStatement("SET SESSION app.attribute TO '" + UUID.randomUUID() + "'").execute())))
+            .last()
+            .subscribe();
+
+        Thread.sleep(2000);
+
+        // Then database should have 15 connection in the pool
+        Integer dbActiveConnectionNumber = Mono.from(getConnectionFactory().create())
+            .flatMap(connection -> Mono.from(connection.createStatement("SELECT count(*) from pg_stat_activity where usename ='james';").execute()))
```
Comment (quantranhong1999) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionPoolTest.java:148]: usename -> username?

Comment (vttranlina) [reply mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionPoolTest.java:148]: `usename`, it is a system table

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java (hunk 3 above)
Comment (vttranlina) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java:142]: `connection` here is `pooledConnection`, one wrapper, 
`.close` method in `pooledConnection` is "release connection to the pool", it does not `close` the physical connection. 

//I same confused when first use it

**Pull request closed without merging**